### ISSUE TITLE
Hex input start key is ignored, so hex input is broken.

### ISSUE THREAD

**Issue by xorgy** [ibus/ibus#1961]
This is a regression from march, caused by: https://github.com/ibus/ibus/commit/02d99aa00cb4a26a2cfb6098ea27e763e9bb2daa#diff-d13f2f35e6ae7938c61d89d377e27184L1058

**Comment by fujiwarat** [comment]
Engine Unicode typing and emoji typing are migrated into panel emoji dialog since engine's shortcut key is hard-coded.

**Comment by xorgy** [comment]
@fujiwarat Could you explain that more? Currently there is no way to enter hex mode _at all_. I'm not touching emoji typing, hex input mode is different entirely. From what I can tell this is a straightforward bug, and has _nothing_ to do with emoji typing.

**Comment by xorgy** [comment]
Hex input mode and the Emoji dialog are _completely different things_. Hex input mode is for inserting arbitrary codepoints by hex number. It is documented on [Wikipedia](https://en.wikipedia.org/wiki/Unicode_input#In_X11_.28Linux_and_Unix_variants.29), the [GNOME documentation](https://help.gnome.org/users/gnome-help/stable/tips-specialchars.html.en#ctrlshiftu). 

The Ctrl + Shift + U keyboard shortcut comes from the ISO/IEC 14755 standard, which is why it is hardcoded. The commit I referenced simply breaks this standard behaviour.

**Comment by xorgy** [comment]
Unabridged related conversation from [#ibus on freenode](irc://irc.freenode.net/#ibus):

> 05:35 < microcolonel> fujiwarat: すみません、藤原さんに https://github.com/ibus/ibus/issues/1961 を勘違いさせちゃえましたかも。説明を伸ばしましたから、もう一度ご覧になり、ご意見をいただければ幸いですが。
> 05:36 < microcolonel> それがバグだと確信しています。
> 05:37 < fujiwarat> microcolonel: 説明しましたように、現在のemoji ui で unicode point も 入力できます。
> 05:41 < microcolonel> But how? The dialog is not letting me do that.
> 05:41 < fujiwarat> For non-GNOME, you can type Ctrl-Shift-e, For GNOME, you can run `ibus emoji` command.
> 05:42 < microcolonel> Yeah, I have the shortcut set up
> 05:43 < microcolonel> Looks like this, and does nothing when I type a codepoint in http://pasteall.org/pic/show.php?id=120416
> 05:44 < microcolonel> Also, it is very slow, it sometimes takes four or five seconds to respond to a keypress.
> 05:44 < fujiwarat> You can type "123" in the text entry for example.
> 05:45 < microcolonel> ohh, it seems like it's a GTK bug causing it not to update the display for some reason
> 05:45 < microcolonel> anyway, that's not the problem
> 05:45 < microcolonel> this shouldn't be in the emoji dialog
> 05:46 < fujiwarat> Do you use the latest ibus 1.5.17? which includes some bug fixes.
> 05:46 < microcolonel> it contradicts the ISO/IEC 14755 standard, the GNOME documentation, Wikipedia, the Archlinux documentation, the Debian documentation, the Chrome OS documentation
> 05:46 < microcolonel> 1.5.17 is the broken one!
> 05:47 < microcolonel> sorry, let me rephrase that. 1.5.17 is when this regression became part of a release.
> 05:47 < fujiwarat> Which desktop do you use? GNOME?
> 05:48 < microcolonel> I use GTK applications outside of GNOME
> 05:49 < fujiwarat> So you run the emoji UI from an ibus panel menu or typing ctrl-shift-e ?
> 05:49 < microcolonel> I run it by pressing the keyboard shortcut
> 05:49 < microcolonel> (by the way, Ctrl Shift E might be a bad idea, it conflicts with popular applications like the GIMP [which uses this for "Export As"])
> 05:50 < microcolonel> I changed it to Ctrl Alt E
> 05:50 < fujiwarat> Yes, that's why ibus-setup provides the shortcut key customization. The default shortcut key is still under the consideration.
> 05:50 < microcolonel> the window updates if I manually move or resize it, but otherwise it freezes, that's just some weird behaviour
> 05:50 < microcolonel> anyway
> 05:50 < microcolonel> the Emoji input dialog is the wrong place for unicode codepoint input
> 05:51 < microcolonel> I type unicode codepoints hundreds of times a day
> 05:51 < microcolonel> the emoji dialog is too invasive
> 05:51 < fujiwarat> microcolonel: We think two shortcut keys ctrl-shift-e and ctrl-shift-u are not needed. The features can be combined.
> 05:52 < microcolonel> All of the code for hex input is still there, and the Ctrl + Shift + U keyboard shortcut is endorsed by an ISO standard, and the documentation of every major desktop environment
> 05:52 < microcolonel> frankly, I should be able to input a unicode codepoint inline without opening a dialog
> 05:52 < microcolonel> I will continue to run this patch on all my machines until it's fixed in upstream, that is how important this is to me.
> 05:54 < fujiwarat> microcolonel: Since ibus-engine's way has the hard-coded shortcut key so it conflicts with other applications so ibus won't have that way.
> 05:54 < microcolonel> hmm
> 05:55 < fujiwarat> Can you see any errors when you run /usr/lib*/ibus-ui-emojier directly ?
> 05:56 < fujiwarat> > Also, it is very slow, it sometimes takes four or five seconds to respond to a keypress.
> 05:56 < microcolonel> I figured that out
> 05:56 < microcolonel> I don't know how to fix it
> 05:56 < microcolonel> but it only updates the contents of the window when X configures it again
> 05:57 < microcolonel> (i.e. by moving it or resizing it)
> 05:57 < microcolonel> I thought it was a delay because I was moving it after about that amount of time
> 05:57 < microcolonel> anyway
> 05:57 < fujiwarat> ? Did you answer my question?
> 05:57 < microcolonel> I can't seem to find it at that path
> 05:57 < microcolonel> lemme see where it's installed for me
> 05:58 < fujiwarat> It's the same path of ibus-ui-gtk3 which should be running.
> 05:58 < microcolonel> yeah, I just configured a different lib path
> 05:58 < microcolonel> /usr/lib/ibus, mmkay
> 05:58 < fujiwarat> Yes, lib*
> 05:59 < microcolonel> the emojier gives some generic warnings, but it runs without error
> 05:59 < microcolonel> and it does copy the character to the clipboard
> 05:59 < microcolonel> but anyway, the dialog is the wrong place for this
> 06:00 < microcolonel> would it be okay if a way were found to make this config-dependent (and maybe add it to ibus-setup)?
> 06:00 < microcolonel> Ctrl Shift U as a default has worked for a very long time
> 06:00 < microcolonel> I don't see why it would suddenly be a problem
> 06:00 < microcolonel> it is an assumption supported by documentation everywhere. GTK does it internally if you don't have ibus
> 06:01 < fujiwarat> Which desktop are you using?
> 06:01 < microcolonel> I just have a window manager as the root
> 06:01 < microcolonel> a fork of dwm
> 06:01 < microcolonel> I run it with xinit
> 06:02 < microcolonel> I don't think there's anything wrong with the hardcoded Ctrl Shift U shortcut, because it is a basic assumption that it will be captured if you are focused in a text input
> 06:03 < microcolonel> nobody could have developed an application which conflicts with this shortcut, because it has been captured for such a long time
> 06:05 < fujiwarat> https://github.com/ibus/ibus/issues/1889
> 06:05  * microcolonel sighs audibly
> 06:05 < microcolonel> that's awful
> 06:05 < fujiwarat> Several applications argue c-s-u.
> 06:05 < microcolonel> they're messy ports though
> 06:06 < microcolonel> some applications use alt click
> 06:06 < microcolonel> but that breaks mutter compatibility!
> 06:07 < fujiwarat> No, I don't think so. You still can setup c-s-u for the emoji ui.
> 06:07 < microcolonel> I don't want the emoji ui
> 06:07 < microcolonel> I want to be able to enter codepoints in hex
> 06:07 < microcolonel> inline
> 06:08 < microcolonel> It's a different use case, in my not so humble opinion.
> 06:08 < fujiwarat> You have a problem something to launch the ui slowly. I'd think it's the way to fix it.
> 06:08 < microcolonel> no it launches just fine
> 06:08 < microcolonel> but it's annoying that it's a window
> 06:08 < microcolonel> because the default behaviour for windows in my environment is to be fullscreen
> 06:08 < microcolonel> so an exception has to be made for it
> 06:08 < microcolonel> and the position it shows up on the screen is random
> 06:09 < microcolonel> and it's broken because I have to move or resize it for the contents to show up
> 06:09 < microcolonel> it's nobody's fault in particular, but the experience is really wrong for my use case
> 06:09 < microcolonel> how about this
> 06:09 < microcolonel> I'll take a crack at making a separate configurable keyboard shortcut
> 06:09 < microcolonel> maybe we can't default it to Ctrl Shift U
> 06:10 < microcolonel> all the code for hex input is still in ibusenginesimple.c
> 06:10 < microcolonel> and it works really really well
> 06:10 < fujiwarat> microcolonel: I think your case is minor which runs full screen. Probably I'd suggest to use ibus-rawcode.
> 06:10 < microcolonel> it is a very valuable feature for me, one that I'm willing to maintain downstream if I have to
> 06:11 < microcolonel> ibus-rawcode has the problem of being another separate input method
> 06:11 < microcolonel> which means that it conflicts with toggling to ibus-kkc
> 06:11 < microcolonel> so I have to press multiple times to get the one I want
> 06:12 < microcolonel> when I type Québec, I'm thinking Shift + Q, u, Ctrl + Shift + U, e, 9, b, e, c
> 06:13 < fujiwarat> microcolonel: Why don't you use compose key features?
> 06:14 < fujiwarat> I don't suppose such a usage.
> 06:14 < microcolonel> because they don't give me enough flexibility
> 06:14 < microcolonel> sometimes I want to type U+303 COMBINING TILDE
> 06:15 < fujiwarat> Multi_key - ' - e outputs é
> 06:15 < microcolonel> or U+201C LEFT DOUBLE QUOTATION MARK, or U+2122 TRADE MARK SIGN
> 06:15 < microcolonel> or U+00B0 DEGREE SIGN
> 06:16 < microcolonel> it's easier to remember the numbers than some arcane set of compose shortcuts
> 06:16 < microcolonel> and those compose chords don't cover the breadth of codepoints I want to use
> 06:17 < fujiwarat> Multi_key - t- m is  U+2122
> 06:18 < microcolonel> No idea how to type a μ but it was natural to me to type [Control + Shift + XK_u, XK_3, XK_b, XK_c] just now, from memory
> 06:20 < microcolonel> anyway, I've gotten used to the ubiquitous behaviour of hex input mode
> 06:20 < fujiwarat> I think most people uses the compose key feature.
> 06:20 < microcolonel> works whether or not the compose key is set up on a machine
> 06:20 < microcolonel> I have never in my life seen somebody use a compose key, but your mileage may vary
> 06:21 < microcolonel> I have no idea how to even get it working anymore
> 06:21 < microcolonel> and I wouldn't really want to, because I need my right alt key
> 06:22 < microcolonel> my base layout (used both for english and for ibus-kkc) is Dvorak as well, so some of the non-mnemonic compose key shortcuts frankly make absolutely no sense 
> 06:22 < fujiwarat> XKB can provides some options for Multi_key besides right alt.
> 06:22 < microcolonel> if I could even get them working
> 06:22 < microcolonel> yeah, I don't have so many keys on my keyboard
> 06:22 < microcolonel> right now I'm using a chromebook
> 06:24 < microcolonel> I guess I could map the extra bottom row key on this ISO/Canadian bilingual layout to Multi_key in latin modes, instead of having it always be 変換
> 06:24 < microcolonel> but anyway, that would force me to figure out how to look up Compose shortcuts, when I already know the codepoints of everything I want, and prefer to type the hex rather than chording my hand
> 06:25 < microcolonel> I get enough key chording from Emacs, I don't want to have to do it just to type ñ
> 06:25 < fujiwarat> I think switching an xkb engine and kkc and switching rawcode and kkc is not so different stress.
> 06:25 < microcolonel> Honestly, there is no way I'm giving up Ctrl + Shift + U
> 06:26 < microcolonel> it is just too valuable to me
> 06:26 < microcolonel> when work forces me to use Windows or OS X to test something, it is one of the main things I miss
> 06:27 < microcolonel> OS X has compose, but it fails subtly
> 06:28 < microcolonel> All of the code is still there in the engine, so if I make it configurable and turn it off by default, will you accept that patch?
> 06:31 < fujiwarat> microcolonel: It's considerable but probably depends on how it can be configurable.
> 06:33 < microcolonel> I mean including a tab in ibus-setup for Hex input
> 06:33 < microcolonel> same shortcut config stuff
> 06:34 < microcolonel> To be honest, it's still a problem if it's not enabled by default though
> 06:34 < microcolonel> because there is a LOT of documentation on the internet which tells users that Ctrl Shift U will let you input hex codepoints inline
> 06:34 < fujiwarat> Probably ibus-engine-simple won't use gsettings.
> 06:34 < microcolonel> ahh
> 06:34 < microcolonel> hmm
> 06:35 < microcolonel> それは困るなぁ
> 06:35 < microcolonel> is there a text config somewhere?
> 06:40 < fujiwarat> microcolonel: OK, probably it's worthwhile for you to try gsettings. 
> 06:42 < microcolonel> hmm
> 06:42 < microcolonel> I'm just thinking, is it even okay not to have it be on by default with Ctrl Shift U though?
> 06:43 < microcolonel> I'm finding a lot of documents telling users that you can use this keyboard shortuct on ibus systems
> 06:43 < microcolonel> that's going to get very, very, very confusing
> 06:44 < microcolonel> Jetbrains users can remap things, but users of every other application can't tell that that documentation no longer applies
> 06:44 < microcolonel> Is somebody going to go edit thousands of documents?
> 06:45 < fujiwarat> I think implementing c-s-u in ibus is not long history and c-s-u is no default for ibus because the hard-coded shortcut key is a bug.
> 06:46 < microcolonel> but not having it on by default is also a bug
> 06:46 < microcolonel> what if it's on by default, but configurable?
> 06:47 < microcolonel> anyway
> 06:47 < fujiwarat> Now it's migrated into c-s-e and the feature is still available.
> 06:47 < microcolonel> Anyhow, 06:40 < fujiwarat> microcolonel: OK, probably it's worthwhile for you to try gsettings. 
> 06:47 < fujiwarat> yes
> 06:48 < microcolonel> ^ by this, do you mean that it would be okay to introduce gsettings in ibus-engine-simple?
> 06:48 < microcolonel> just want to be clear before I go do any work
> 06:51 < fujiwarat> Yes. ibus-engine-simple and gtk-im-context-simple are almost same. I thought if ibus-engine-simple will have the gsettings, we cannot sync the feature to gtk-im-context-simple so I have thought not to have gsettings in i-e-s but probably it's good for me to reconsider it.
> 06:53 < microcolonel> okay, I'll go make that happen, and we can decide later if it's possible to make it default on again. ;- )
> 06:53 < fujiwarat> Probably I also think the settings won't touch ibus-setup.
> 06:55 < microcolonel> Maybe I should do this in two stages, the first one just a boolean setting to enable or disable the feature
> 06:56 < microcolonel> then make it configurable in the second stage
> 06:57 < microcolonel> actually, where does the emoji shortcut from ibus-setup get saved?
> 06:57 < microcolonel> I don't see it anywhere in /desktop/ibus in dconf
> 06:57 < fujiwarat> If just a boolean, I think an environment variable might be an option, IBUS_ENABLE_CTRL_SHIFT_U.
> 06:58 < microcolonel> oh, it's in /desktop/ibus/panel instead of /desktop/ibus/general
> 06:58 < microcolonel> hmm
> 06:58 < fujiwarat> I mean to type gsettings command directly.
> 06:59 < microcolonel> how do you mean?
> 07:00 < fujiwarat> E.g. gsettings list-keys org.freedesktop.ibus.general.hotkey
> 07:00 < microcolonel> yeah
> 07:01 < microcolonel> the emoji hotkey is stored in org.freedesktop.ibus.panel.emoji.hotkey
> 07:01 < microcolonel> it's in the gtk_accelerator_parse format though
> 07:01 < microcolonel> probably don't want to add gtk calls to ibusenginesimple
> 07:02 < microcolonel> so that's out of the question
> 07:02 < microcolonel> boolean is probably the only way
> 07:02 < fujiwarat> Right. I separated the schemas since ibus has several emoji settings.
> 07:02 < microcolonel> I'm thinking a boolean to toggle whether hex mode is on is probably the best way to go
> 07:03 < fujiwarat> `gsettings set` and `gsettings get` can change an value.
> 07:03 < microcolonel> right, but the hotkey is in a format that has to be parsed
> 07:04 < microcolonel> it's a string which contains what looks like an array literal and a sequence of tokens
> 07:04 < microcolonel> it's in a format recognized by a gtk function, gtk_accelerator_parse
> 07:04 < microcolonel> it seems like a LOT of work to reimplement that in ibusenginesimple.ca
> 07:04 < microcolonel> *.c
> 07:05 < fujiwarat> So I dropped the feature in ibus-enigne-simple.
> 07:05 < microcolonel> I'll just do boolean for now then
> 07:08 < fujiwarat> I think an environment variable in ibus-engine-simple or adding an argument for ibus-daemon works for you?
> 07:09 < microcolonel> hmm, environment variable it is
> 07:09 < fujiwarat> Probably I think gsettings is not needed for the boolean settings at the moment.
> 07:09 < microcolonel> well, just from a user perspective
> 07:10 < microcolonel> I figure it should be more accessible
> 07:10 < microcolonel> if not on by default, it should at least be straightforward for a casual user to turn it on.
> 07:10 < microcolonel> in my opinion.
> 07:11 < fujiwarat> I'd like to leave the user perspective until other users argue to drop c-s-u by default.
> 07:14 < fujiwarat> Probably I think it would be better to paste this chat in the bug report.

**Comment by fujiwarat** [comment]
The updated patch sounds good to me at the moment.
